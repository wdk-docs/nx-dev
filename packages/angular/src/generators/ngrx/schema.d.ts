export interface NgRxGeneratorOptions {
  directory: string;
  minimal: boolean;
  module: string;
  name: string;
  useDataPersistence: boolean;
  barrels?: boolean;
  facade?: boolean;
  root?: boolean;
  skipFormat?: boolean;
  skipImport?: boolean;
  skipPackageJson?: boolean;
  syntax?: 'classes' | 'creators';
}
