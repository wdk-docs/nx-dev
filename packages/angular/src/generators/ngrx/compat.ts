import { convertNxGenerator } from '@nrwl/devkit';
import { ngrxGenerator } from './ngrx';

export default convertNxGenerator(ngrxGenerator);
