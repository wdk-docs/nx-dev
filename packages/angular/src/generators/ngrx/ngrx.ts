import type { GeneratorCallback, Tree } from '@nrwl/devkit';
import { formatFiles } from '@nrwl/devkit';
import {
  addExportsToBarrel,
  addImportsToModule,
  addNgRxToPackageJson,
  generateNgrxFilesFromTemplates,
  normalizeOptions,
} from './lib';
import type { NgRxGeneratorOptions } from './schema';

export async function ngrxGenerator(
  tree: Tree,
  options: NgRxGeneratorOptions
): Promise<GeneratorCallback> {
  const normalizedOptions = normalizeOptions(options);

  if (!tree.exists(normalizedOptions.module)) {
    throw new Error(`Module does not exist: ${normalizedOptions.module}.`);
  }

  if (!normalizedOptions.minimal || !normalizedOptions.root) {
    generateNgrxFilesFromTemplates(tree, normalizedOptions);
  }

  if (!normalizedOptions.skipImport) {
    addImportsToModule(tree, normalizedOptions);
    addExportsToBarrel(tree, normalizedOptions);
  }

  let packageInstallationTask: GeneratorCallback = () => {};
  if (!normalizedOptions.skipPackageJson) {
    packageInstallationTask = addNgRxToPackageJson(tree);
  }

  if (!normalizedOptions.skipFormat) {
    await formatFiles(tree);
  }

  return packageInstallationTask;
}

export default ngrxGenerator;
