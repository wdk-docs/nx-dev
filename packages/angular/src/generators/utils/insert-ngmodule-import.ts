import { applyChangesToString, ChangeType, Tree } from '@nrwl/devkit';
import {
  __String,
  CallExpression,
  createSourceFile,
  ImportDeclaration,
  isArrayLiteralExpression,
  isCallExpression,
  isClassDeclaration,
  isIdentifier,
  isImportDeclaration,
  isNamedImports,
  isObjectLiteralExpression,
  isPropertyAssignment,
  ObjectLiteralExpression,
  PropertyAssignment,
  ScriptTarget,
  SourceFile,
} from 'typescript';

type ngModuleDecoratorProperty =
  | 'imports'
  | 'providers'
  | 'declarations'
  | 'exports';

export function insertNgModuleProperty(
  tree: Tree,
  modulePath: string,
  name: string,
  property: ngModuleDecoratorProperty
) {
  const contents = tree.read(modulePath).toString('utf-8');

  const sourceFile = createSourceFile(
    modulePath,
    contents,
    ScriptTarget.ESNext
  );

  const coreImport = findImport(sourceFile, '@angular/core');

  if (!coreImport) {
    throw new Error(
      `There are no imports from "@angular/core" in ${modulePath}.`
    );
  }

  const ngModuleNamedImport = getNamedImport(coreImport, 'NgModule');

  const ngModuleName = ngModuleNamedImport.name.escapedText;

  const ngModuleClassDeclaration = findDecoratedClass(sourceFile, ngModuleName);

  const ngModuleDecorator = ngModuleClassDeclaration.decorators.find(
    (decorator) =>
      isCallExpression(decorator.expression) &&
      isIdentifier(decorator.expression.expression) &&
      decorator.expression.expression.escapedText === ngModuleName
  );

  const ngModuleCall = ngModuleDecorator.expression as CallExpression;

  if (ngModuleCall.arguments.length < 1) {
    const newContents = applyChangesToString(contents, [
      {
        type: ChangeType.Insert,
        index: ngModuleCall.getEnd() - 1,
        text: `{ ${property}: [${name}]}`,
      },
    ]);
    tree.write(modulePath, newContents);
  } else {
    if (!isObjectLiteralExpression(ngModuleCall.arguments[0])) {
      throw new Error(
        `The NgModule options for ${ngModuleClassDeclaration.name.escapedText} in ${modulePath} is not an object literal`
      );
    }

    const ngModuleOptions = ngModuleCall
      .arguments[0] as ObjectLiteralExpression;

    const typeProperty = findPropertyAssignment(ngModuleOptions, property);

    if (!typeProperty) {
      let text = `${property}: [${name}]`;
      if (ngModuleOptions.properties.hasTrailingComma) {
        text = `${text},`;
      } else {
        text = `, ${text}`;
      }
      const newContents = applyChangesToString(contents, [
        {
          type: ChangeType.Insert,
          index: ngModuleOptions.getEnd() - 1,
          text,
        },
      ]);
      tree.write(modulePath, newContents);
    } else {
      if (!isArrayLiteralExpression(typeProperty.initializer)) {
        throw new Error(
          `The NgModule ${property} for ${ngModuleClassDeclaration.name.escapedText} in ${modulePath} is not an array literal`
        );
      }

      let text: string;
      if (typeProperty.initializer.elements.hasTrailingComma) {
        text = `${name},`;
      } else {
        text = `, ${name}`;
      }
      const newContents = applyChangesToString(contents, [
        {
          type: ChangeType.Insert,
          index: typeProperty.initializer.getEnd() - 1,
          text,
        },
      ]);
      tree.write(modulePath, newContents);
    }
  }
}

export function insertNgModuleImport(
  tree: Tree,
  modulePath: string,
  importName: string
) {
  insertNgModuleProperty(tree, modulePath, importName, 'imports');
}

function findImport(sourceFile: SourceFile, importPath: string) {
  const importStatements = sourceFile.statements.filter(isImportDeclaration);

  return importStatements.find(
    (statement) =>
      statement.moduleSpecifier
        .getText(sourceFile)
        .replace(/['"`]/g, '')
        .trim() === importPath
  );
}

function getNamedImport(coreImport: ImportDeclaration, importName: string) {
  if (!isNamedImports(coreImport.importClause.namedBindings)) {
    throw new Error(
      `The import from ${coreImport.moduleSpecifier} does not have named imports.`
    );
  }

  return coreImport.importClause.namedBindings.elements.find((namedImport) =>
    namedImport.propertyName
      ? isIdentifier(namedImport.propertyName) &&
        namedImport.propertyName.escapedText === importName
      : isIdentifier(namedImport.name) &&
        namedImport.name.escapedText === importName
  );
}

function findDecoratedClass(sourceFile: SourceFile, ngModuleName: __String) {
  const classDeclarations = sourceFile.statements.filter(isClassDeclaration);
  return classDeclarations.find(
    (declaration) =>
      declaration.decorators &&
      declaration.decorators.some(
        (decorator) =>
          isCallExpression(decorator.expression) &&
          isIdentifier(decorator.expression.expression) &&
          decorator.expression.expression.escapedText === ngModuleName
      )
  );
}

function findPropertyAssignment(
  ngModuleOptions: ObjectLiteralExpression,
  propertyName: ngModuleDecoratorProperty
) {
  return ngModuleOptions.properties.find(
    (property) =>
      isPropertyAssignment(property) &&
      isIdentifier(property.name) &&
      property.name.escapedText === propertyName
  ) as PropertyAssignment;
}
