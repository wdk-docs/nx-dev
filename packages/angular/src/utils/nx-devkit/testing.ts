import type { Tree } from '@nrwl/devkit';
import { names, updateJson } from '@nrwl/devkit';

export interface AppConfig {
  appName: string; // name of app
  appModule: string; // app/app.module.ts in the above sourceDir
}
export interface LibConfig {
  name: string;
  module: string;
  barrel: string;
}

var appConfig: AppConfig; // configure built in createApp()
var libConfig: LibConfig;

export function getAppConfig(): AppConfig {
  return appConfig;
}
export function getLibConfig(): LibConfig {
  return libConfig;
}

export function createApp(
  tree: Tree,
  appName: string,
  routing: boolean = true
) {
  appName = names(appName).fileName;
  // save for getAppDir() lookup by external *.spec.ts tests
  appConfig = {
    appName,
    appModule: `/apps/${appName}/src/app/app.module.ts`,
  };

  tree.write(
    appConfig.appModule,
    `
     import { NgModule } from '@angular/core';
     import { BrowserModule } from '@angular/platform-browser';
     ${routing ? "import { RouterModule } from '@angular/router'" : ''};
     import { AppComponent } from './app.component';
     @NgModule({
       imports: [BrowserModule, ${routing ? 'RouterModule.forRoot([])' : ''}],
       declarations: [AppComponent],
       bootstrap: [AppComponent]
     })
     export class AppModule {}
  `
  );
  tree.write(
    `/apps/${appName}/src/main.ts`,
    `
    import { enableProdMode } from '@angular/core';
    import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';

    import { AppModule } from './app/app.module';
    import { environment } from './environments/environment';

    if (environment.production) {
      enableProdMode();
    }

    platformBrowserDynamic()
      .bootstrapModule(AppModule)
      .catch(err => console.log(err));
  `
  );
  tree.write(
    `/apps/${appName}/tsconfig.app.json`,
    JSON.stringify({
      include: ['**/*.ts'],
    })
  );
  tree.write(
    `/apps/${appName}-e2e/tsconfig.e2e.json`,
    JSON.stringify({
      include: ['../**/*.ts'],
    })
  );
  tree.write(
    '/workspace.json',
    JSON.stringify({
      newProjectRoot: '',
      version: 1,
      projects: {
        [appName]: {
          root: `apps/${appName}`,
          sourceRoot: `apps/${appName}/src`,
          architect: {
            build: {
              options: {
                main: `apps/${appName}/src/main.ts`,
              },
            },
            serve: {
              options: {},
            },
          },
          tags: [],
        },
      },
    })
  );
}

export function createLib(tree: Tree, libName: string) {
  const { name, className, fileName, propertyName } = names(libName);

  libConfig = {
    name,
    module: `/libs/${propertyName}/src/lib/${fileName}.module.ts`,
    barrel: `/libs/${propertyName}/src/index.ts`,
  };

  tree.write(
    libConfig.module,
    `
      import { NgModule } from '@angular/core';
      import { CommonModule } from '@angular/common';
      @NgModule({
        imports: [
          CommonModule
        ],
        providers: []
      })
      export class ${className}Module { }
  `
  );
  tree.write(
    libConfig.barrel,
    `
    export * from './lib/${fileName}.module';
  `
  );
}
