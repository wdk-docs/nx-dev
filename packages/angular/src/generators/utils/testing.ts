import type { Tree } from '@nrwl/devkit';
import { wrapAngularDevkitSchematic } from '@nrwl/devkit/ngcli-adapter';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { Linter } from '@nrwl/linter';
import { UnitTestRunner } from '../../utils/test-runners';
import libraryGenerator from '../library/library';

export async function createStorybookTestWorkspaceForLib(
  libName: string
): Promise<Tree> {
  let tree = createTreeWithEmptyWorkspace();

  const moduleGenerator = wrapAngularDevkitSchematic(
    '@schematics/angular',
    'module'
  );
  const componentGenerator = wrapAngularDevkitSchematic(
    '@schematics/angular',
    'component'
  );

  await libraryGenerator(tree, {
    name: libName,
    buildable: false,
    linter: Linter.EsLint,
    publishable: false,
    simpleModuleName: false,
    skipFormat: false,
    unitTestRunner: UnitTestRunner.Jest,
  });

  await componentGenerator(tree, {
    name: 'test-button',
    project: libName,
  });

  tree.write(
    `libs/${libName}/src/lib/test-button/test-button.component.ts`,
    `import { Component, Input } from '@angular/core';

export type ButtonStyle = 'default' | 'primary' | 'accent';

@Component({
  selector: 'proj-test-button',
  templateUrl: './test-button.component.html',
  styleUrls: ['./test-button.component.css']
})
export class TestButtonComponent {
  @Input('buttonType') type = 'button';
  @Input() style: ButtonStyle = 'default';
  @Input() age?: number;
  @Input() isOn = false;
}`
  );

  tree.write(
    `libs/${libName}/src/lib/test-button/test-button.component.html`,
    `<button [attr.type]="type" [ngClass]="style"></button>`
  );

  const modulePath = `libs/${libName}/src/lib/${libName}.module.ts`;
  tree.write(
    modulePath,
    `import * as ButtonExports from './test-button/test-button.component';
    ${tree.read(modulePath)}`
  );

  // create a module with component that gets exported in a barrel file
  await moduleGenerator(tree, {
    name: 'barrel',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'barrel-button',
    project: libName,
    path: `libs/${libName}/src/lib/barrel`,
    module: 'barrel',
  });

  tree.write(
    `libs/${libName}/src/lib/barrel/barrel-button/index.ts`,
    `export * from './barrel-button.component';`
  );

  tree.write(
    `libs/${libName}/src/lib/barrel/barrel.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BarrelButtonComponent } from './barrel-button';

@NgModule({
  imports: [CommonModule],
  declarations: [BarrelButtonComponent],
})
export class BarrelModule {}`
  );

  // create a module with components that get Angular exported and declared by variable
  await moduleGenerator(tree, {
    name: 'variable-declare',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'variable-declare-button',
    project: libName,
    path: `libs/${libName}/src/lib/variable-declare`,
    module: 'variable-declare',
  });

  await componentGenerator(tree, {
    name: 'variable-declare-view',
    project: libName,
    path: `libs/${libName}/src/lib/variable-declare`,
    module: 'variable-declare',
  });

  tree.write(
    `libs/${libName}/src/lib/variable-declare/variable-declare.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VariableDeclareButtonComponent } from './variable-declare-button/variable-declare-button.component';
import { VariableDeclareViewComponent } from './variable-declare-view/variable-declare-view.component';

const COMPONENTS = [
  VariableDeclareButtonComponent,
  VariableDeclareViewComponent
]

@NgModule({
  imports: [CommonModule],
  declarations: COMPONENTS,
  exports: COMPONENTS
})
export class VariableDeclareModule {}`
  );

  // create a module with components that get Angular exported and declared by variable
  await moduleGenerator(tree, {
    name: 'variable-spread-declare',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-button',
    project: libName,
    path: `libs/${libName}/src/lib/variable-spread-declare`,
    module: 'variable-spread-declare',
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-view',
    project: libName,
    path: `libs/${libName}/src/lib/variable-spread-declare`,
    module: 'variable-spread-declare',
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-anotherview',
    project: libName,
    path: `libs/${libName}/src/lib/variable-spread-declare`,
    module: 'variable-spread-declare',
  });

  tree.write(
    `libs/${libName}/src/lib/variable-spread-declare/variable-spread-declare.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VariableSpreadDeclareButtonComponent } from './variable-spread-declare-button/variable-spread-declare-button.component';
import { VariableSpreadDeclareViewComponent } from './variable-spread-declare-view/variable-spread-declare-view.component';
import { VariableSpreadDeclareAnotherviewComponent } from './variable-spread-declare-anotherview/variable-spread-declare-anotherview.component';

const COMPONENTS = [ 
  VariableSpreadDeclareButtonComponent, 
  VariableSpreadDeclareViewComponent 
]

@NgModule({
  imports: [CommonModule],
  declarations: [...COMPONENTS, VariableSpreadDeclareAnotherviewComponent],
})
export class VariableSpreadDeclareModule {}`
  );

  // create a module where declared components are pulled from a static member of the module
  await moduleGenerator(tree, {
    name: 'static-member-declarations',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'cmp1',
    project: libName,
    path: `libs/${libName}/src/lib/static-member-declarations`,
    module: 'static-member-declarations',
  });

  await componentGenerator(tree, {
    name: 'cmp2',
    project: libName,
    path: `libs/${libName}/src/lib/static-member-declarations`,
    module: 'static-member-declarations',
  });

  tree.write(
    `libs/${libName}/src/lib/static-member-declarations/static-member-declarations.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Cmp1Component } from './cmp1/cmp1.component';
import { Cmp2Component } from './cmp2/cmp2.component';

@NgModule({
  imports: [CommonModule],
  declarations: StaticMemberDeclarationsModule.COMPONENTS,
  exports: StaticMemberDeclarationsModule.COMPONENTS
})
export class StaticMemberDeclarationsModule {
  static readonly COMPONENTS = [Cmp1Component, Cmp2Component];
}`
  );

  // create another button in a nested subpath
  await moduleGenerator(tree, {
    name: 'nested',
    project: libName,
    path: `libs/${libName}/src/lib`,
  });

  await componentGenerator(tree, {
    name: 'nested-button',
    project: libName,
    module: 'nested',
    path: `libs/${libName}/src/lib/nested`,
  });

  await componentGenerator(tree, {
    name: 'test-other',
    project: libName,
  });

  return tree;
}
