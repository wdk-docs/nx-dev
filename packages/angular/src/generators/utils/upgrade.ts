import type { GeneratorCallback, Tree } from '@nrwl/devkit';
import { installPackagesTask, updateJson } from '@nrwl/devkit';
import { angularJsVersion } from '../../utils/versions';

export function addUpgradeToPackageJson(tree: Tree): GeneratorCallback {
  updateJson(tree, './package.json', (json) => {
    json.dependencies = {
      '@angular/upgrade': json.dependencies['@angular/core'],
      angular: angularJsVersion,
      ...(json.dependencies ?? {}),
    };
    json.dependencies = sortObjectByKeys(json.dependencies);

    return json;
  });

  return () => {
    installPackagesTask(tree);
  };
}

function sortObjectByKeys(obj: unknown): unknown {
  return Object.keys(obj)
    .sort()
    .reduce((result, key) => {
      return {
        ...result,
        [key]: obj[key],
      };
    }, {});
}
