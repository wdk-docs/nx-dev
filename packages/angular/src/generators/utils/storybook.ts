import type { Tree } from '@nrwl/devkit';
import {
  findNodes,
  getSourceNodes,
} from '@nrwl/workspace/src/utilities/typescript';
import type { PropertyDeclaration } from 'typescript';
import { SyntaxKind } from 'typescript';
import { getTsSourceFile } from '../../utils/nx-devkit/ast-utils';

export type KnobType = 'text' | 'boolean' | 'number' | 'select';
export interface InputDescriptor {
  name: string;
  type: KnobType;
  defaultValue?: string;
}

export function getInputPropertyDeclarations(
  tree: Tree,
  path: string
): PropertyDeclaration[] {
  const file = getTsSourceFile(tree, path);

  const decorators = getSourceNodes(file).filter(
    (node) => node.kind === SyntaxKind.Decorator
  );

  return decorators
    .filter((decorator) =>
      findNodes(decorator, SyntaxKind.Identifier).some(
        (node) => node.getText() === 'Input'
      )
    )
    .map((node) => node.parent as PropertyDeclaration);
}

export function getComponentProps(
  tree: Tree,
  componentPath: string,
  getArgsDefaultValueFn: (property: PropertyDeclaration) => string | undefined
): InputDescriptor[] {
  const props = getInputPropertyDeclarations(tree, componentPath).map(
    (node) => {
      const decoratorContent = findNodes(
        findNodes(node, SyntaxKind.Decorator)[0],
        SyntaxKind.StringLiteral
      );
      const name = decoratorContent.length
        ? decoratorContent[0].getText().slice(1, -1)
        : node.name.getText();

      const type = getKnobType(node);
      const defaultValue = getArgsDefaultValueFn(node);

      return {
        name,
        type,
        defaultValue,
      };
    }
  );

  return props;
}

export function getKnobType(property: PropertyDeclaration): KnobType {
  if (property.type) {
    const typeName = property.type.getText();
    const typeNameToKnobType: Record<string, KnobType> = {
      string: 'text',
      number: 'number',
      boolean: 'boolean',
    };
    return typeNameToKnobType[typeName] || 'text';
  }
  if (property.initializer) {
    const initializerKindToKnobType: Record<number, KnobType> = {
      [SyntaxKind.StringLiteral]: 'text',
      [SyntaxKind.NumericLiteral]: 'number',
      [SyntaxKind.TrueKeyword]: 'boolean',
      [SyntaxKind.FalseKeyword]: 'boolean',
    };
    return initializerKindToKnobType[property.initializer.kind] || 'text';
  }
  return 'text';
}
