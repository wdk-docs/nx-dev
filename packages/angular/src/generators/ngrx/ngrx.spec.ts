import type { Tree } from '@nrwl/devkit';
import * as devkit from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { createApp } from '../../utils/nx-devkit/testing';
import { ngrxVersion } from '../../utils/versions';
import { ngrxGenerator } from './ngrx';
import type { NgRxGeneratorOptions } from './schema';

describe('ngrx', () => {
  let tree: Tree;
  const defaultOptions: NgRxGeneratorOptions = {
    directory: '+state',
    minimal: true,
    module: 'apps/myapp/src/app/app.module.ts',
    name: 'users',
    useDataPersistence: false,
  };

  const expectFileToExist = (file: string) =>
    expect(tree.exists(file)).toBeTruthy();

  beforeEach(() => {
    jest.clearAllMocks();
    tree = createTreeWithEmptyWorkspace();
    createApp(tree, 'myapp');
  });

  it('should error when the module could not be found', async () => {
    const modulePath = 'not-existing.module.ts';

    await expect(
      ngrxGenerator(tree, {
        ...defaultOptions,
        module: modulePath,
      })
    ).rejects.toThrowError(`Module does not exist: ${modulePath}.`);
  });

  it('should add an empty root module when minimal and root are set to true', async () => {
    await ngrxGenerator(tree, {
      ...defaultOptions,
      root: true,
      minimal: true,
    });

    expect(
      tree.read('/apps/myapp/src/app/app.module.ts', 'utf-8')
    ).toMatchSnapshot();
  });

  it('should not generate files when minimal and root are set to true', async () => {
    await ngrxGenerator(tree, {
      ...defaultOptions,
      root: true,
      minimal: true,
    });

    expect(tree.exists('/apps/myapp/src/app/+state/users.actions.ts')).toBe(
      false
    );
    expect(tree.exists('/apps/myapp/src/app/+state/users.effects.ts')).toBe(
      false
    );
    expect(
      tree.exists('/apps/myapp/src/app/+state/users.effects.spec.ts')
    ).toBe(false);
    expect(tree.exists('/apps/myapp/src/app/+state/users.reducer.ts')).toBe(
      false
    );
    expect(tree.exists('/apps/myapp/src/app/+state/users.selectors.ts')).toBe(
      false
    );
    expect(
      tree.exists('/apps/myapp/src/app/+state/users.selectors.spec.ts')
    ).toBe(false);
  });

  it('should add a root module with feature module when minimal is set to false', async () => {
    await ngrxGenerator(tree, {
      ...defaultOptions,
      root: true,
      minimal: false,
    });

    expect(
      tree.read('/apps/myapp/src/app/app.module.ts', 'utf-8')
    ).toMatchSnapshot();
  });

  it('should not add RouterStoreModule when the module does not reference the router', async () => {
    createApp(tree, 'no-router-app', false);

    await ngrxGenerator(tree, {
      ...defaultOptions,
      module: 'apps/no-router-app/src/app/app.module.ts',
      root: true,
    });

    const appModule = tree.read(
      '/apps/no-router-app/src/app/app.module.ts',
      'utf-8'
    );
    expect(appModule).not.toContain('StoreRouterConnectingModule.forRoot()');
  });

  it('should add facade provider when facade is true', async () => {
    await ngrxGenerator(tree, {
      ...defaultOptions,
      root: true,
      minimal: false,
      facade: true,
    });

    expect(tree.read('/apps/myapp/src/app/app.module.ts', 'utf-8')).toContain(
      'providers: [UsersFacade]'
    );
  });

  it('should not add facade provider when facade is false', async () => {
    await ngrxGenerator(tree, {
      ...defaultOptions,
      root: true,
      minimal: false,
      facade: false,
    });

    expect(
      tree.read('/apps/myapp/src/app/app.module.ts', 'utf-8')
    ).not.toContain('providers: [UsersFacade]');
  });

  it('should not add facade provider when minimal is true', async () => {
    await ngrxGenerator(tree, {
      ...defaultOptions,
      root: true,
      minimal: true,
      facade: true,
    });

    expect(
      tree.read('/apps/myapp/src/app/app.module.ts', 'utf-8')
    ).not.toContain('providers: [UsersFacade]');
  });

  it('should not generate imports when skipImport is true', async () => {
    await ngrxGenerator(tree, {
      ...defaultOptions,
      minimal: false,
      skipImport: true,
    });

    expectFileToExist('/apps/myapp/src/app/+state/users.actions.ts');
    expectFileToExist('/apps/myapp/src/app/+state/users.effects.ts');
    expectFileToExist('/apps/myapp/src/app/+state/users.effects.spec.ts');
    expectFileToExist('/apps/myapp/src/app/+state/users.reducer.ts');
    expectFileToExist('/apps/myapp/src/app/+state/users.selectors.ts');
    expectFileToExist('/apps/myapp/src/app/+state/users.selectors.spec.ts');
    expect(
      tree.read('/apps/myapp/src/app/app.module.ts', 'utf-8')
    ).toMatchSnapshot();
  });

  it('should update package.json', async () => {
    await ngrxGenerator(tree, defaultOptions);

    const packageJson = devkit.readJson(tree, 'package.json');
    expect(packageJson.dependencies['@ngrx/store']).toEqual(ngrxVersion);
    expect(packageJson.dependencies['@ngrx/effects']).toEqual(ngrxVersion);
    expect(packageJson.dependencies['@ngrx/entity']).toEqual(ngrxVersion);
    expect(packageJson.dependencies['@ngrx/router-store']).toEqual(ngrxVersion);
    expect(packageJson.dependencies['@ngrx/component-store']).toEqual(
      ngrxVersion
    );
    expect(packageJson.devDependencies['@ngrx/schematics']).toEqual(
      ngrxVersion
    );
    expect(packageJson.devDependencies['@ngrx/store-devtools']).toEqual(
      ngrxVersion
    );
    expect(packageJson.devDependencies['jasmine-marbles']).toBeDefined();
  });

  it('should not update package.json when skipPackageJson is true', async () => {
    await ngrxGenerator(tree, { ...defaultOptions, skipPackageJson: true });

    const packageJson = devkit.readJson(tree, 'package.json');
    expect(packageJson.dependencies['@ngrx/store']).toBeUndefined();
    expect(packageJson.dependencies['@ngrx/effects']).toBeUndefined();
    expect(packageJson.dependencies['@ngrx/entity']).toBeUndefined();
    expect(packageJson.dependencies['@ngrx/router-store']).toBeUndefined();
    expect(packageJson.dependencies['@ngrx/component-store']).toBeUndefined();
    expect(packageJson.devDependencies['@ngrx/schematics']).toBeUndefined();
    expect(packageJson.devDependencies['@ngrx/store-devtools']).toBeUndefined();
  });

  it('should format files', async () => {
    jest.spyOn(devkit, 'formatFiles');

    await ngrxGenerator(tree, defaultOptions);

    expect(devkit.formatFiles).toHaveBeenCalled();
  });

  it('should not format files when skipFormat is true', async () => {
    jest.spyOn(devkit, 'formatFiles');

    await ngrxGenerator(tree, { ...defaultOptions, skipFormat: true });

    expect(devkit.formatFiles).not.toHaveBeenCalled();
  });
});
