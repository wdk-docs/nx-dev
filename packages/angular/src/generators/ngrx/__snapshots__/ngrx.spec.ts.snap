// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ngrx should add a root module with feature module when minimal is set to false 1`] = `
"
     import { NgModule } from '@angular/core';
     import { BrowserModule } from '@angular/platform-browser';
     import { RouterModule } from '@angular/router';
     import { AppComponent } from './app.component';
import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';
import * as fromUsers from './+state/users.reducer';
import { UsersEffects } from './+state/users.effects';
import { NxModule } from '@nrwl/angular';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';
import { environment } from '../environments/environment';
import { StoreRouterConnectingModule } from '@ngrx/router-store';
     @NgModule({
       imports: [BrowserModule, RouterModule.forRoot([]), NxModule.forRoot(), StoreModule.forRoot({}, {
      metaReducers: !environment.production ? [] : [],
      runtimeChecks: {
        strictActionImmutability: true,
        strictStateImmutability: true
      }
    }), EffectsModule.forRoot([UsersEffects]), !environment.production ? StoreDevtoolsModule.instrument() : [], StoreRouterConnectingModule.forRoot(), StoreModule.forFeature(fromUsers.USERS_FEATURE_KEY, fromUsers.reducer)],
       declarations: [AppComponent],
       bootstrap: [AppComponent]
     })
     export class AppModule {}
  "
`;

exports[`ngrx should add an empty root module when minimal and root are set to true 1`] = `
"
     import { NgModule } from '@angular/core';
     import { BrowserModule } from '@angular/platform-browser';
     import { RouterModule } from '@angular/router';
     import { AppComponent } from './app.component';
import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';
import { environment } from '../environments/environment';
import { StoreRouterConnectingModule } from '@ngrx/router-store';
     @NgModule({
       imports: [BrowserModule, RouterModule.forRoot([]), StoreModule.forRoot({}, {
      metaReducers: !environment.production ? [] : [],
      runtimeChecks: {
        strictActionImmutability: true,
        strictStateImmutability: true
      }
    }), EffectsModule.forRoot([]), !environment.production ? StoreDevtoolsModule.instrument() : [], StoreRouterConnectingModule.forRoot()],
       declarations: [AppComponent],
       bootstrap: [AppComponent]
     })
     export class AppModule {}
  "
`;

exports[`ngrx should not generate imports when skipImport is true 1`] = `
"
     import { NgModule } from '@angular/core';
     import { BrowserModule } from '@angular/platform-browser';
     import { RouterModule } from '@angular/router';
     import { AppComponent } from './app.component';
     @NgModule({
       imports: [BrowserModule, RouterModule.forRoot([])],
       declarations: [AppComponent],
       bootstrap: [AppComponent]
     })
     export class AppModule {}
  "
`;
